components/nifosg/nifloader.cpp

```diff
@@ -1289,50 +1280,81 @@ namespace NifOsg
                         std::string filename = Misc::ResourceHelpers::correctTexturePath(st->filename, textureManager->getVFS());
 
                         unsigned int clamp = static_cast<unsigned int>(tex.clamp);
-                        int wrapT = (clamp) & 0x1;
-                        int wrapS = (clamp >> 1) & 0x1;
+                        bool wrapT = (clamp) & 0x1;
+                        bool wrapS = (clamp >> 1) & 0x1;
 
-                        osg::Texture2D* texture2d = textureManager->getTexture2D(filename,
+                        osg::ref_ptr<osg::Texture2D> texture2d = textureManager->getTexture2D(filename,
                               wrapS ? osg::Texture::REPEAT : osg::Texture::CLAMP,
                               wrapT ? osg::Texture::REPEAT : osg::Texture::CLAMP);
 
                         int texUnit = boundTextures.size();
 
-                        stateset->setTextureAttributeAndModes(texUnit, texture2d, osg::StateAttribute::ON);
-
-                        if (i == Nif::NiTexturingProperty::GlowTexture)
+                        osg::ref_ptr<osg::TexEnv> texEnv = NULL;
+                        switch(i)
                         {
-                            osg::TexEnvCombine* texEnv = new osg::TexEnvCombine;
-                            texEnv->setCombine_Alpha(osg::TexEnvCombine::REPLACE);
-                            texEnv->setSource0_Alpha(osg::TexEnvCombine::PREVIOUS);
-                            texEnv->setCombine_RGB(osg::TexEnvCombine::ADD);
-                            texEnv->setSource0_RGB(osg::TexEnvCombine::PREVIOUS);
-                            texEnv->setSource1_RGB(osg::TexEnvCombine::TEXTURE);
-
-                            stateset->setTextureAttributeAndModes(texUnit, texEnv, osg::StateAttribute::ON);
+                        case Nif::NiTexturingProperty::BaseTexture:
+                            texEnv = new osg::TexEnv;
+                            break;
+                        case Nif::NiTexturingProperty::GlowTexture:
+                        {
+                            osg::TexEnvCombine* texEnvCombine = new osg::TexEnvCombine;
+                            texEnvCombine->setCombine_Alpha(osg::TexEnvCombine::REPLACE);
+                            texEnvCombine->setSource0_Alpha(osg::TexEnvCombine::PREVIOUS);
+                            texEnvCombine->setCombine_RGB(osg::TexEnvCombine::ADD);
+                            texEnvCombine->setSource0_RGB(osg::TexEnvCombine::PREVIOUS);
+                            texEnvCombine->setSource1_RGB(osg::TexEnvCombine::TEXTURE);
+
+                            texEnv = dynamic_cast<osg::TexEnv*>(texEnvCombine);
```

Comment:
This cast will always fail, because TexEnvCombine is not derived from TexEnv.

I really don't like where this iteration process is going. Can we please just focus on the one thing we want to fix, and leave the rest of the code alone? I suggest to start from scratch, then replace the if (i != Nif::NiTexturingProperty::BaseTexture...) statement with a switch statement, that logs a different error message for each texturing type. That is the only code you need to change.